File: dsa/patterns/05_binarysearch/python/l410_split_array_largest_sum.py
# LeetCode 410 - Split Array Largest Sum
# (Book Allocation / Painter's Partition variant)
#
# Idea:
# Binary search on the answer.
# - Minimum possible max sum = max(nums)
# - Maximum possible max sum = sum(nums)
#
# Check feasibility by counting how many subarrays
# (students) are needed for a given max sum.
#
# Time Complexity: O(n log(sum(nums)))
# Space Complexity: O(1)

def can_split(nums: list[int], k, max_sum: int) -> int:
    sub_arrays = 1
    current_sum = 0

    for num in nums:
        if current_sum + num <= max_sum:
            current_sum += num
        else:
            sub_arrays += 1
            current_sum = num

    return sub_arrays <= k


def split_array(nums: list[int], k: int) -> int:
    # Impossible to split
    if k > len(nums):
        return -1

    left, right = max(nums), sum(nums)
    ans = right

    while left <= right:
        mid = (left + right) // 2

        if can_split(nums, k, mid):
            ans = mid
            right = mid - 1   # try to minimize further
        else:
            left = mid + 1

    return ans


# Driver code
if __name__ == "__main__":
    nums = [25, 46, 28, 49, 24]
    m = 4
    print(split_array(nums, m))  # 71

File: dsa/patterns/05_binarysearch/java/L410_SplitArrayLargestSum.java
import java.util.Collections;
import java.util.List;

public class L410_SplitArrayLargestSum {

	public static void main(String[] args) {
		L410_SplitArrayLargestSum obj = new L410_SplitArrayLargestSum();

		int[] nums = new int[] {25, 46, 28, 49, 24};
		System.out.println(obj.findPages(nums, 4));
	}

	// Problem: https://leetcode.com/problems/split-array-largest-sum/
    // Minimize the max pages allocated to k students
    // Minimize the max sum of array after splitting array into k parts
    // Idea: Apply BS on range max and sum.
    public int findPages(int[] nums, int m) {
        int n = nums.length;
        // book allocation impossible
        if (m > n)
            return -1;

        int max = Integer.MIN_VALUE, sum = 0;
        for(int num: nums){
            max = Math.max(max, num);
            sum += num;
        }

        int low = max, high = sum, ans = sum;
        while (low <= high) {
            int mid = (low + high) / 2;
            int students = countStudents(nums, mid);
            if (students <= m) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }

    public static int countStudents(int[] nums, int pages) {
        int n = nums.length;
        int students = 1;
        long pagesStudent = 0;
        for (int i = 0; i < n; i++) {
            if (pagesStudent + nums[i] <= pages) {
                pagesStudent += nums[i];
            } else {
                students++;
                pagesStudent = nums[i];
            }
        }

        return students;
    }
}
